### ISSUE TITLE
Rendering list of metadata tags in results

### ISSUE THREAD

**Issue by jimbul** [microsoft-search/pnp-modern-search#85]
**Version used**
Ex: 3.9

**Describe the bug**
Search results condense all metadata tags together into a single string for each result item with no commas, semicolons or spaces between the tags.

**To Reproduce**
1. Do a search for an item tagged with multiple managed metadata tags.
2. Look at results.

**Expected behavior**
Metadata tags to be separated by spaces, commas or semicolons.

**Screenshots**

<img width="457" alt="Screenshot 2020-02-03 at 11 10 22 AM" src="https://user-images.githubusercontent.com/4074243/73622898-d06ace80-4675-11ea-8456-30643efc5ed8.png">

**Desktop (please complete the following information):**
 Same issue seen on all mainstream browsers, safari, firefox and chrome.

**Comment by PooLP** [comment]
@FranckyC : I see this and other design problem this week, if you want.

**Comment by FranckyC** [comment]
Hi @PooLP sure, if you have a fix, feel free!

**Comment by phillipharding** [comment]
You can use the built-in helpers to achieve the same effect

```RefinableString106``` is mapped to a multi-value taxonomy field in the example below;

```
<div class="post-tags">
    {{#each (split RefinableString106 ";") as |tag| }}
        <span class="post-tag">
            <i class="ms-Icon ms-Icon--MapPin" aria-hidden="true"></i>
            <a href="?q=RefinableString106:'{{trim tag}}'" data-interception="off">{{tag}}</a>
        </span>
    {{/each}}
</div>
```

**Comment by joelfmrodrigues** [comment]
@PooLP Are you going to look into this issue? If not, I may submit a PR

**Comment by FranckyC** [comment]
@joelfmrodrigues @PooLP  First PR wins ;). Thanks @phillipharding for the answer !

**Comment by PooLP** [comment]
@joelfmrodrigues : ah yes. thanks for PR ;)

**Comment by FranckyC** [comment]
@jimbul Does the answer given by @phillipharding fix your issue?

**Comment by jimbul** [comment]
Hi @FranckyC - I'm not sure - where would i paste @phillipharding's suggestion? in the results template?

Thanks for the responses everyone.

**Comment by FranckyC** [comment]
Hi @jimbul , you must put this in your Handlebars template. You have the same example where to put this in the 'Simple List' layout:

```
<div class="template_tags example-themePrimary">
	{{#if owstaxidmetadataalltagsinfo}}
		<i class="ms-Icon ms-Icon--Tag" aria-hidden="true"></i>
		{{#each (split owstaxidmetadataalltagsinfo ",") as |tag| }}                                                    
			<a href="#owstaxidmetadataalltagsinfo:'{{trim tag}}'">{{tag}}</a>
		{{/each}}
	{{/if}}
</div>
```

**Comment by FranckyC** [comment]
I mark this one as resolved since a valid solution has been provided.

**Comment by jimbul** [comment]
Hi All - i updated the template for simple list as instructed, but now see no tags at all so have reverted and will wait for the solution to be included in a release. My understanding of how this works is the issue here - i appreciate your support.

**Comment by FranckyC** [comment]
Hi @jimbul . There is no fix for this since it is not an issue per say. What managed property are you using to display tags?

**Comment by jimbul** [comment]
Hi @FranckyC  - thanks for your reply - i'm using the simple list results layout - and i'm not consciously displaying the tags they are just showing up under the results. we're using RefinableString00-07. What am i misunderstanding - thanks again!

**Comment by FranckyC** [comment]
The default simple list layout use the `owstaxidmetadataalltagsinfo` default managed property to display tags. Can you post a screenshot of you see?

**Comment by jimbul** [comment]
sure @FranckyC: here you go - this is with standard list view - if i add the code above, (which i did not edit at all - and realise i do not have a refinablestring106), i see no tags at all.
![image](https://user-images.githubusercontent.com/4074243/74794512-5afe2f80-52fe-11ea-976f-0fb2f15b894d.png)

**Comment by FranckyC** [comment]
The default template already includes this code snippet so no need to add it by yourself. However, you can just replace the default `owstaxidmetadataalltagsinfo` property with the one you use for multi taxonomy values (like `REfinableStringXX`).

**Comment by FranckyC** [comment]
@jimbul does this resolve your issue?

**Comment by wobba** [comment]
Closing as inactive.

**Comment by smitman3405** [comment]
I added the code above and I still see the "garbley-goop" instead of just the Tag Text.
L0|#06fd7161c-fa8f-403e-a8d6-62e33d8709f4|Collaboration Services;L0|#0cef9b086-0ba6-4a24-a9d3-c856299d7092|Azure;L0|#0c8b95930-f1e0-4601-bfb3-7b9e66405a45|Manage;L0|#07c3bf0ff-1c0e-4bbc-8fbd-e754d01b44be|Microsoft;L0|#0af6033f2-fb5e-47a9-875a-6ffbfc5ad2e0|Project Management;L0|#03c095735-1ca1-4547-9d03-6b1b5fc8d4c5|Sales
So basically I need to change the template, split on "|" instead of "," but I only want the LAST text in the "tag" object, but I don't know the syntax for it.  I tried: 
{{#each (last ( split owstaxidmetadataalltagsinfo "|")) as |tag| }} 
but that didn't work.

![image](https://user-images.githubusercontent.com/10451710/107237157-6d341a80-69f4-11eb-8bb5-d587d38271e8.png)

**Comment by FranckyC** [comment]
@smitman3405 You have to enable localization in the search results Web Part settings to see only the label.

**Comment by smitman3405** [comment]
Thank you for the quick reply!

Enable localization AND make the change to the template?
I modified the webpart and reset the template to the default template text and I'm still seeing the extra info around the taxonomy items.
![image](https://user-images.githubusercontent.com/10451710/107244933-948ee580-69fc-11eb-8c52-92e2bd996fec.png)

**Comment by FranckyC** [comment]
Are you using v4.0?

**Comment by smitman3405** [comment]
yes:
![image](https://user-images.githubusercontent.com/10451710/107245653-62ca4e80-69fd-11eb-884a-0ebc5d4c0949.png)

**Comment by FranckyC** [comment]
In addition to the localization setting, in the "slots" section, you must prefix your managed property with the term `Auto`

**Comment by smitman3405** [comment]
I'm not sure I understand your instructions.
In this image I'm showing the "slots" section, do I need to rename the Tags "slot" or create a new one for each like owstaxIdServicex0020Tags with a name like Auto-Service-Tags?
![image](https://user-images.githubusercontent.com/10451710/107249262-30baeb80-6a01-11eb-89a4-aeb9c3292653.png)

**Comment by FranckyC** [comment]
Prefix `owsTaxIdServicex0020Tags` with `Auto` so `AutoowsTaxIdServicex0020Tags`

**Comment by smitman3405** [comment]
again, not fully understanding what it is you want me to try, but I tried this:
and it did not work.
![image](https://user-images.githubusercontent.com/10451710/107254247-20a50b00-6a05-11eb-91bb-259a91cffdce.png)

I was expecting a change to the template something like this, where I'm displaying the 2nd (zero based) item in tags array from the split
                        <div class="template--listItem--tags example-themePrimary">
                            {{#if (slot item @root.slots.Tags)}}
                            <pnp-icon data-name="Tag" aria-hidden="true" data-theme-variant="{{JSONstringify @root.theme}}"></pnp-icon>
                            {{#each (split (slot item @root.slots.Tags) ",") as |tag| }}
                            <a href="#{{slot item @root.slots.Tags}}:'{{trim tag}}'">{{tag**[2]**}}</a> {{/each}} {{/if}}
                        </div>

**Comment by smitman3405** [comment]
I've got it to work with the following template change:
						<div class="template_tags example-themePrimary">
							{{#if owstaxidmetadataalltagsinfo}}
								<pnp-icon data-name="Tag" aria-hidden="true" data-theme-variant="{{JSONstringify @root.theme}}"></pnp-icon>
								{{#each (split owstaxidmetadataalltagsinfo ";") as |tag| }}                                                    
									<a href="https://<mysite>.sharepoint.com/sites/KB/SitePages/Metadata-Search.aspx?q='{{last (split tag "|")}}'">{{last (split tag "|")}}</a>
								{{/each}}
							{{/if}}
						</div>

the problem I'm having is the first {{last (split tag "|")}} is returning the entire metadata item:
<a href="https://<mysite>.sharepoint.com/sites/KB/SitePages/Metadata-Search.aspx?q='L0|#059549cc6-c31b-403e-9c2e-b68cbb370bef|Cloud Management Services'">Cloud Management Services</a> instead of just the last array in the tag object.

if I can't reference the last item in the object in the href - then how do I store that value in a variable in handlebar syntax.

**Comment by FranckyC** [comment]
@smitman3405 I mean this value should be set as the slot value, not the display name.

**Comment by smitman3405** [comment]
I have many taxonomy items and they may change over time.  It doesn't seem flexible if I have to code each taxonomy item as a slot.  the code I reference above is working, except for the first {{last (split tag "|")}} is returning the entire tag value and not just the last item.
how do I set a variable to the value in the last object of tag?

**Comment by FranckyC** [comment]
@smitman3405 you don't have to 'code' each taxonomy item as a slot. Using the `Auto` prefix will automatically translate all your term values automatically for the managed property. You don't need to update the template to extract value.

**Comment by smitman3405** [comment]
I've created the following Slots:
![image](https://user-images.githubusercontent.com/10451710/107360201-9b246800-6aa3-11eb-9224-31961c939576.png)
using the predefined template the results look like this:
![image](https://user-images.githubusercontent.com/10451710/107360346-ca3ad980-6aa3-11eb-8175-857a8bc6aae5.png)

what am I missing?

**Comment by meany** [comment]
@smitman3405 and anyone else looking for this solution. I have a field in my library called "Tags", and this is where I set the value of `Autoowstaxidmetadataalltagsinfo`.

![image](https://user-images.githubusercontent.com/854996/113881163-2eb7a380-978a-11eb-8aa0-836f017dfd7b.png)

**Comment by smitman3405** [comment]
I was able to use the tag value by setting up the tag fields as Search Properties in SharePoint Admin and mapping the fields to the RefinableStringXX fields.  It took a bit of time for them to show but then I was able to map the DisplayName to the FilterField with the RefinableStringXX field.



 

 





I hope this helps

 

Mike-

**Comment by sanazubair-debug** [comment]
Please help  me find out, multiline choice column managed property are mapped and refiners are mapped, the values which are comming with multine choice column is with semicolon and hashtag, both in refiners and search result webpart. I need to replace it with comma and space after each value in search result. 
And in refiners with mutliline choice refinable string 128,should not come with hashtag

**Comment by Charan-Thota** [comment]
@sanazubair-debug No sure if this is is what you want but after updating udpating the tags to **Autoowstaxidmetadataalltagsinfo**, I updated the template with this code to acheive below results,

Code:
`{{#each (split (slot item @root.slots.Tags) "|") as |tags| }}
{{#each (split tags ",") as |tag| }}
		<pnp-icon data-name="Tag" aria-hidden="true" data-theme-variant="{{JSONstringify @root.theme}}"></pnp-icon>
                 <a href="#'{{trim tag}}'">{{trim tag }}</a> 
{{/each}}
{{/each}}`

result changed from this 
![image](https://user-images.githubusercontent.com/26864248/124212934-25449a00-db33-11eb-819c-f7c39be88ad7.png)

to 

![image](https://user-images.githubusercontent.com/26864248/124212736-d4cd3c80-db32-11eb-83a5-79ac668fcec0.png)

**Comment by corriespondent** [comment]
> You can use the built-in helpers to achieve the same effect
> 
> `RefinableString106` is mapped to a multi-value taxonomy field in the example below;
> 
> ```
> <div class="post-tags">
>     {{#each (split RefinableString106 ";") as |tag| }}
>         <span class="post-tag">
>             <i class="ms-Icon ms-Icon--MapPin" aria-hidden="true"></i>
>             <a href="?q=RefinableString106:'{{trim tag}}'" data-interception="off">{{tag}}</a>
>         </span>
>     {{/each}}
> </div>
> ```

I set up my search box to use the dynamic query from the query parameter "q," but I get 0 results when I try this method. Is it not possible to use refinement filters via a query string?

I also tried the other method with <a href="#owstaxidmetadataalltagsinfo:'{{trim tag}}'">{{tag}}</a> as described by @FranckyC, but clicking the tags don't DO anything. Am I missing some other configuration that would allow that filter to work?
